tron/utils/scribereader.py

```diff
@@ -0,0 +1,158 @@
+    with open("/nail/etc/ecosystem") as f:
+        return f.read().strip()
+
+
+@lru_cache(maxsize=1)
+def get_scribereader_host_and_port() -> Optional[Tuple[str, int]]:
+    try:
+        ecosystem = get_ecosystem()
+        superregion = get_superregion()
+        region = get_region()
+    except OSError:
+        log.warning("Unable to read location mapping files from disk, not returning scribereader host/port")
+        return None
+
+    host, port = scribereader.get_tail_host_and_port(
+        ecosystem=ecosystem if ecosystem != "prod" else None, region=region, superregion=superregion,
+    )
+    return host, port
+
+
+def read_log_stream_for_action_run(
+    action_run_id: str, component: str, min_date: Optional[datetime.datetime], max_date: Optional[datetime.datetime],
+) -> List[str]:
+    if min_date is None:
+        return [f"{action_run_id} has not started yet."]
+
+    if scribereader is None:
+        return ["Scribereader (an internal Yelp package) is not available - unable to display logs."]
+    if get_scribereader_host_and_port() is None:
+        return [
+            "Unable to determine where Tron is located. If you're seeing this inside Yelp, report this to #compute-infra"
+        ]
+
+    host, port = get_scribereader_host_and_port()  # type: ignore  # the None case is covered by the check above
+
+    today = datetime.date.today()
+    start_date = min_date.date()
+    end_date = max_date.date() if max_date else None
+
+    use_tailer = today in {start_date, end_date}
+    use_reader = start_date != today and end_date is not None
+
+    if end_date is not None and end_date == today:
+        end_date -= datetime.timedelta(days=1)
+
+    namespace, job_name, run_num, action = action_run_id.split(".")
+    # in our logging infra, things are logged to per-instance streams - but
+    # since Tron PaaSTA instances are of the form `job_name.action`, we need
+    # to escape the period since some parts of our infra will reject streams
+    # containing them - thus, the "weird" __ separator
+    stream_name = f"stream_paasta_app_output_{namespace}_{job_name}__{action}"
+    output: List[Tuple[str, str]] = []
+
+    # We'll only use a stream reader for logs from not-today.
+    # that said, it's possible that an action spans more than a single day - in this case, we'll first read "historical" data from
+    # the reader and then follow-up with today's logs from a stream tailer.
+    # NOTE: this is more-or-less what our internal `scribereader` binary does
+    if use_reader:
+        with scribereader.get_stream_reader(
+            stream_name=stream_name, min_date=min_date, max_date=max_date, reader_host=host, reader_port=port,
+        ) as stream:
+            for line in stream:
+                try:
+                    payload = json.loads(line)
+                except json.decoder.JSONDecodeError:
+                    log.error(f"Unable to decode log line from stream ({stream_name}) for {action_run_id}: {line}")
```

Comment:
nit: I wonder if this is something that we should report in the output somehow, like whether or not there were a lot of lines that couldn't be decoded, but it's not strictly necessary.

Reply:
hopefully this never actually gets hit :p 

that said, that's a good idea and I can add a line to the output somewhere with the count of undecodable lines pretty easily